### ISSUE TITLE
0011686:
Editing Users / Sync with Active Directory doesn't work

### ISSUE THREAD

**Issue by Gloirin** [tine20/Tine-2.0-Open-Source-Groupware-and-CRM#5782]
**Reported by errorsmith on 11 Mar 2016 18:00**

**Version:** git master

After editing a user in the webinterface and clicking on &quot;save&quot;, the application throws the following error:

0x43 (Operation not allowed on RDN; 00002016: Modify of RDN &#039;CN&#039; on CN=Albert Einstein,OU=internal,OU=user,DC=domain,DC=intra not permitted, must use &#039;rename&#039; operation instead): updating: CN=Albert Einstein,OU=internal,OU=user,DC=domain,DC=intra
.../Tinebase/User/ActiveDirectory.php(349): Zend_Ldap-&gt;update()
.../Tinebase/User/Sql.php(759): Tinebase_User_ActiveDirectory-&gt;updateUserInSyncBackend()
.../Admin/Controller/User.php(205): Tinebase_User_Sql-&gt;updateUser()
.../Admin/Frontend/Json.php(374): Admin_Controller_User-&gt;update()
[internal function]: Admin_Frontend_Json-&gt;saveUser()
.../library/Zend/Server/Abstract.php(232): call_user_func_array()
.../Zend/Json/Server.php(558): Zend_Server_Abstract-&gt;_dispatch()
.../Zend/Json/Server.php(197): Zend_Json_Server-&gt;_handle()
.../Tinebase/Server/Json.php(293): Zend_Json_Server-&gt;handle()
.../Tinebase/Server/Json.php(166): Tinebase_Server_Json-&gt;_handle()
.../Tinebase/Core.php(229): Tinebase_Server_Json-&gt;handle()
.../index.php(18): Tinebase_Core::dispatchRequest()

This happens even if one opens the &quot;edit user&quot; page and presses &quot;save without changing anything. Newly created users doesn&#039;t appear in the OU configured in tine-setup but somewhere else. Additionally, after editing a user directly in the Active Directory, the changes / not all changes aren&#039;t reflected in tine 2.0. 

This results in tine 2.0 being not usable with an active directory as account storage.

**Steps to reproduce:** 1.: Set up a server with Active directory 

2.: Create the neccessary users for the initial configuration

3.: configure tine 2.0 for authentication via ldap and account storage active directory

4.a: create a user in tine 2.0, search it in AD, find it somewhere else than configured the configured OU.

4.b create a user in AD. Find it in tine 2.0 as expected. Modify the user in AD (i.e. add user to a group). Find the changes aren&#039;t reflected in tine 2.0

4.c try to add a user (same as step 4.b or another, previously untouched user) to a group in tine 2.0. This results in the above mentioned error

**Additional information:** My main configuration consists of an samba4 active directory, running on FreeBSD 10.2. I&#039;m using apache 2.4 / php 5.6 on as a webserver and mariadb 10.1 for the mysql database. To exclude errors from samba, I set up a Windows 2008R2 for testing, but this resulted the same errors as with samba, so I suspect there&#039;s a problem with tine 2.0 itself.

**Comment by Gloirin** [comment]
**Comment posted by pschuele on 15 Mar 2016 12:41**

as suggested in the forum (https://www.tine20.org/forum/viewtopic.php?f=12&amp;t=17906&amp;p=63982#p63982), you should try the &quot;RFC2307bis&quot; option in the setup.

maybe this should be the default for the AD backend config?

**Comment by Gloirin** [comment]
**Comment posted by pschuele on 15 Mar 2016 13:05**

ah, it does not seem to be a RFC2307bis-issue. it seems that the cn ist updated somehow which is not permitted.

I&#039;ll have a look at this in my test environment.

**Comment by Gloirin** [comment]
**Comment posted by errorsmith on 15 Mar 2016 13:22**

Thanks for the reply. I will proceed with this discussion in the forum and carry over any results over here.

**Comment by Gloirin** [comment]
**Comment posted by pschuele on 30 Mar 2016 18:30**

just uploaded a patch (created with git format-patch -1).

please have a look.

it fixed the problems for me...

**Comment by Gloirin** [comment]
**Comment posted by errorsmith on 30 Mar 2016 20:23**

Hi
I unpacked &quot;elena&quot; in a seperate directory and applied the patch with the command:
&quot;git apply --check 0001-0011686-Editing-Users-Sync-with-Active-Directory-doe.patch&quot;

After that, I uploaded the modified copy of elena to the webspace and did a new setup (browser -&gt; domain.de/ti/setup.php) with a new, clean database.

During the setup I get again the error:

0x43 (Operation not allowed on RDN; 00002016: Modify of RDN &#039;CN&#039; on CN=Domain Users,OU=groups,DC=domain,DC=intra not permitted, must use &#039;rename&#039; operation instead): updating: CN=Domain Users,OU=groups,DC=domain,DC=intra
.../Tinebase/Group/ActiveDirectory.php(319): Zend_Ldap-&gt;update()
.../Tinebase/Group/Sql.php(459): Tinebase_Group_ActiveDirectory-&gt;updateGroupInSyncBackend()
.../Tinebase/Group.php(233): Tinebase_Group_Sql-&gt;updateGroup()
.../Tinebase/Group.php(183): Tinebase_Group::syncListsOfUserContact()
.../Tinebase/User.php(511): Tinebase_Group::syncMemberships()
.../Tinebase/User.php(723): Tinebase_User::syncUser()
.../Addressbook/Setup/Initialize.php(38): Tinebase_User::syncUsers()
.../Setup/Initialize.php(44): Addressbook_Setup_Initialize-&gt;_initialize()
.../Setup/Controller.php(1489): Setup_Initialize::initialize()
.../Setup/Controller.php(1383): Setup_Controller-&gt;_installApplication()
.../Setup/Controller.php(918): Setup_Controller-&gt;installApplications()
.../Setup/Frontend/Json.php(241): Setup_Controller-&gt;saveAuthentication()
[internal function]: Setup_Frontend_Json-&gt;saveAuthentication()
.../library/Zend/Server/Abstract.php(232): call_user_func_array()
.../Zend/Json/Server.php(558): Zend_Server_Abstract-&gt;_dispatch()
.../Zend/Json/Server.php(197): Zend_Json_Server-&gt;_handle()
.../Setup/Server/Json.php(90): Zend_Json_Server-&gt;handle()
.../Setup/Core.php(109): Setup_Server_Json-&gt;handle()
.../setup.php(14): Setup_Core::dispatchRequest()

Looks like it isn&#039;t fixed at all places. I will test with a current copy from git later this day. If requested, I can test with egon too. Actually I would like to see this fixed in Egon, so I can set up my installation with the current release. 

edit: I needed some time to find out how to apply the patch. My first answer reflected this. Now I found out and chenged my answer because of this.

Kind regards,
errorsmith

**Comment by Gloirin** [comment]
**Comment posted by pschuele on 31 Mar 2016 07:11**

http://gerrit.tine20.com/customers/#/c/2983/

**Comment by Gloirin** [comment]
**Comment posted by pschuele on 31 Mar 2016 07:36**

hi,

are you sure you updated Tinebase/Group/ActiveDirectory.php, too?

I&#039;ll attach a patched version of the file...

i just did a fresh installation without any errors.

**Comment by Gloirin** [comment]
**Comment posted by errorsmith on 31 Mar 2016 07:43**

To be honest: I&#039;m not sure if the patch applied completely. I executed the above command and got no error. I&#039;ll try the modified file later this day. Just to be sure: Do I need to use elena or egon to test it?

**Comment by Gloirin** [comment]
**Comment posted by pschuele on 31 Mar 2016 08:22**

&gt;  Just to be sure: Do I need to use elena or egon to test it?

they should both work as nothing has changed in the AD classes since Elena :)

**Comment by Gloirin** [comment]
**Comment posted by pschuele on 31 Mar 2016 08:25**

just uploaded a zip with both relevant patched files.

**Comment by Gloirin** [comment]
**Comment posted by errorsmith on 31 Mar 2016 17:25**

Hi

As you said &quot;both relevant patched files&quot; - The .patch references to a third file (somewhere inside the Adressbook subdirectory). Is this file not needed to be patched? Anyway, I tried with a new installation using egon. For this, I unpacked the tine20-allinone_2016.03.1.tar.bz2 and copied the twi files from your .zip into the relevant subdirectories of the Tinebase application. 

Unfortunately I&#039;m still unable to finnish setup. But this time the error changed:

0x20 (No such object; 00002030: No such Base DN: CN=Domain Users,OU=groups,DC=domain,DC=intra): updating: CN=Domain Users,OU=groups,DC=domain,DC=intra
.../Tinebase/Group/ActiveDirectory.php(330): Zend_Ldap-&gt;update()
.../Tinebase/Group/Sql.php(459): Tinebase_Group_ActiveDirectory-&gt;updateGroupInSyncBackend()
.../Tinebase/Group.php(233): Tinebase_Group_Sql-&gt;updateGroup()
.../Tinebase/Group.php(183): Tinebase_Group::syncListsOfUserContact()
.../Tinebase/User.php(511): Tinebase_Group::syncMemberships()
.../Tinebase/User.php(723): Tinebase_User::syncUser()
.../Addressbook/Setup/Initialize.php(38): Tinebase_User::syncUsers()
.../Setup/Initialize.php(44): Addressbook_Setup_Initialize-&gt;_initialize()
.../Setup/Controller.php(1489): Setup_Initialize::initialize()
.../Setup/Controller.php(1383): Setup_Controller-&gt;_installApplication()
.../Setup/Controller.php(918): Setup_Controller-&gt;installApplications()
.../Setup/Frontend/Json.php(241): Setup_Controller-&gt;saveAuthentication()
[internal function]: Setup_Frontend_Json-&gt;saveAuthentication()
.../library/Zend/Server/Abstract.php(232): call_user_func_array()
.../Zend/Json/Server.php(558): Zend_Server_Abstract-&gt;_dispatch()
.../Zend/Json/Server.php(197): Zend_Json_Server-&gt;_handle()
.../Setup/Server/Json.php(90): Zend_Json_Server-&gt;handle()
.../Setup/Core.php(109): Setup_Server_Json-&gt;handle()
.../setup.php(14): Setup_Core::dispatchRequest()

To make things worse, setup did some changes to my AD I did not approve:

1. It moved the group &quot;Domain Users&quot; from &quot;CN=Domain Users,OU=groups,DC=domain,DC=intra&quot; to &quot;CN=Domain Users,OU=user,DC=domain,DC=intra&quot;
The group &quot;Domain Users&quot; definitively existed in OU=groups before starting setup. I double-checked this, moved it back to the groups-OU and, after deleting the database, rerun setup with the same result. After moving the group it complained about the group not beeing where it should be. 

2. It created an additional Tine 2.0 Admin account as &quot;CN=Tine 2.0 Admin Account,OU=user,DC=domain,DC=intra&quot;. This makes no sense, as there already was an admin-account created before as &quot;CN=tine-admin,....&quot;. This account was given to setup when configuring the authorization/accountstorage and should have been used by setup instead of creating a new account.

with kind regards,
errorsmith

**Comment by Gloirin** [comment]
**Comment posted by errorsmith on 31 Mar 2016 18:18**

Some additions:

I somehow managed to login as tine-admin and as user and found the following additional problems (might be related to the above problems):


- again, not all users were syncronized from AD during setup. 
I only found 3 of 10 users in tine 2.0.

- the user role doesn&#039;t have the necessary permission to run most applications 
(projects is allowed for some reasons, but addressbook not etc). The admin role isn&#039;t able to get adressbook running despite having the permissions. Adressbook just isn&#039;t available.

- the group &quot;Domain Users&quot; gets moved back into &quot;ou=user,...&quot; whenever something initiates a sync from AD. It doesn&#039;t matter if the sync is done manually (setup.php) or i.e. by doing some administrative task.

- I&#039;ve got a notification asking me to update to Egon despite the fact that I&#039;m running a fresh installation of Egon everytime I&#039;m loggin in as tine-admin.

- Adding a user to a group results in an error-message: &quot;Sorry, your request could not be completed because the required data could not be found. In most cases this means that someone already deleted the data. Please refresh your current view.&quot;

kind regards,
errorsmith

**Comment by Gloirin** [comment]
**Comment posted by pschuele on 1 Apr 2016 07:52**

hi,

hm, I did not expect so many issues with the patch ... :(

&gt; As you said &quot;both relevant patched files&quot; - The .patch references to a third file 
&gt; (somewhere inside the Adressbook subdirectory). Is this file not needed to be 
&gt; patched?

nope, that&#039;s just a cosmetic change. just replace the two Tinebase/../ActiveDirectory.php files.

&gt; 0x20 (No such object; 00002030: No such Base DN: CN=Domain 
&gt; Users,OU=groups,DC=domain,DC=intra): updating: CN=Domain 
&gt; Users,OU=groups,DC=domain,DC=intra

Which baseDn did you configure for groups? Is it still OU=groups,DC=domain,DC=intra?

I think, this might be a separate issue. It looks like the group baseDn is not used. I&#039;ll verify that. we should open a new bug for that.

&gt; ... might be related to the above problems

yes, i think so.

except for this one:

&gt; - I&#039;ve got a notification asking me to update to Egon despite the fact that I&#039;m running 
&gt; a fresh installation of Egon everytime I&#039;m loggin in as tine-admin.

I&#039;ll check that as well.

**Comment by Gloirin** [comment]
**Comment posted by errorsmith on 1 Apr 2016 08:03**

Hi

&gt;nope, that&#039;s just a cosmetic change. just replace the two Tinebase/../ActiveDirectory.php files.

I did exactly that. 

&gt;Which baseDn did you configure for groups? Is it still OU=groups,DC=domain,DC=intra?

Yes. I doublechecked this and rerun setup twice (each time with a fresh DB) to exclude configuration errors. 
I can make a new bugreport today evening. Currently I&#039;m not able to get to the machine where tine 2.0 is running and therefore I can not attach any logs etc. When I make a report, I&#039;d rather like to make a qualified report containing all relevant information.

Regarding the update-notification: It&#039;s only annoying and not a show-stopper like the issue we&#039;re currently investigating. I can make a bug-report for this too if desired.

kind regards,
errorsmith

**Comment by Gloirin** [comment]
**Comment posted by errorsmith on 4 Apr 2016 16:59**

Hi

I&#039;m now testing the proposed modification to file Tinebase/Group/ActiveDirectory.php:
I&#039;ve changed line 319 to 

$newDn = &quot;cn={$ldapData[&#039;cn&#039;]},{$this-&gt;_options[&#039;groupsDn&#039;]}&quot;;

So far it looks promising, Ive completed setup without errors and I&#039;m now in the progress of testing everything.

Edit: 
I found something not working as expected:
The synchronization from LDAP doesn&#039;t work correctly after editing a user in AD.

Steps to reproduce:
1. create a user (in tine or in AD doesn&#039;t matter) and assign the user to groups. (&quot;Domain Users&quot; &amp; &quot;internal&quot; in my case
2. login as the user, do something, logout
3. modify user directly in AD using the server-manager or RSAT tools: remove the group assignment, give the user other groups. (I moved it from internal to external users, I didn&#039;t touch &quot;Domain Users&quot;)
4. login as tine admin, find the user looks unmodified
5. force sync from ldap using the command &quot;php setup.php --verbose --sync_accounts_from_ldap&quot;
6. login as tine-admin -&gt; user is still unmodified.

Question regarding this:
Does sync_accounts_from_ldap check for useraccounts to sync only in UserDN or does it check the OU recursively? It does so during intitial setup, but I&#039;m not sure about the later sync process.

i.e.: 
my UserDN is in OU=users,DC=domain,DC=intra 
Inside the OU=users exist several OU&#039;s for internal users, external users and so on.
The user-account itself is located in CN=user name,OU=external,OU=users,DC=domain,DC=intra

kind regards,
errorsmith

**Comment by Gloirin** [comment]
**Comment posted by pschuele on 5 Apr 2016 14:04**

hi errorsmith,

I&#039;ll create a new issue for the &quot;new&quot; synchronization problem.

Philipp
